fix sanity test failures

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->


**What happened**:
follow:  https://github.com/csi-driver/blobfuse-csi-driver/tree/master/test/sanity

You will get following sanity test failures:
```
Summarizing 6 Failures:

[Fail] Controller Service CreateVolume [It] should not fail when requesting to create a volume with already existing name and same capacity.
/root/go/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:371

[Fail] Controller Service CreateVolume [It] should fail when requesting to create a volume with already existing name and different capacity.
/root/go/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:413

[Fail] Controller Service CreateVolume [It] should not fail when creating volume with maximum-length name
/root/go/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:491

[Fail] Controller Service DeleteVolume [It] should return appropriate values (no optional values added)
/root/go/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:570

[Fail] Controller Service ValidateVolumeCapabilities [It] should return appropriate values (no optional values added)
/root/go/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:642

[Fail] Node Service [It] should work
/root/go/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/node.go:375

Ran 26 of 57 Specs in 1.100 seconds
FAIL! -- 20 Passed | 6 Failed | 0 Pending | 31 Skipped
--- FAIL: TestSanity (1.11s)
```

**What you expected to happen**:

**How to reproduce it**:

**Anything else we need to know?**:

**Environment**:
- CSI Driver version:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

```
Summarizing 3 Failures:
[Fail] Controller Service CreateVolume [It] should not fail when requesting to create a volume with already existing name and same capacity. 
/home/travis/gopath/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:372
[Fail] Controller Service CreateVolume [It] should fail when requesting to create a volume with already existing name and different capacity. 
/home/travis/gopath/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:443
[Fail] Node Service [It] should work 
/home/travis/gopath/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/node.go:741
Ran 26 of 61 Specs in 29.945 seconds
FAIL! -- 23 Passed | 3 Failed | 0 Pending | 35 Skipped
--- FAIL: TestSanity (29.97s)
```

```
Summarizing 2 Failures:

[Fail] Controller Service CreateVolume [It] should not fail when requesting to create a volume with already existing name and same capacity.
/root/go/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:372

[Fail] Controller Service CreateVolume [It] should fail when requesting to create a volume with already existing name and different capacity.
/root/go/src/github.com/csi-driver/blobfuse-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:443

Ran 26 of 61 Specs in 2.317 seconds
FAIL! -- 24 Passed | 2 Failed | 0 Pending | 35 Skipped
--- FAIL: TestSanity (2.33s)

```

--- a/pkg/blobfuse/blobfuse.go
+++ b/pkg/blobfuse/blobfuse.go
@@ -22,9 +22,11 @@ import (
 
 	"github.com/container-storage-interface/spec/lib/go/csi"
 	csicommon "github.com/csi-driver/blobfuse-csi-driver/pkg/csi-common"
+	"github.com/pborman/uuid"
 
 	"k8s.io/klog"
 	"k8s.io/kubernetes/pkg/util/mount"
+	k8sutil "k8s.io/kubernetes/pkg/volume/util"
 	"k8s.io/legacy-cloud-providers/azure"
 )
 
@@ -39,6 +41,10 @@ const (
 	defaultFileMode  = "0777"
 	defaultDirMode   = "0777"
 	defaultVers      = "3.0"
+
+	// See https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names
+	containerNameMinLength = 3
+	containerNameMaxLength = 63
 )
 
 // Driver implements all interfaces of CSI drivers
@@ -180,4 +186,39 @@ func getStorageAccount(secrets map[string]string) (string, string, error) {
 	}
 
 	return accountName, accountKey, nil
+}
+
+// A container name must be a valid DNS name, conforming to the following naming rules:
+//	1. Container names must start with a letter or number, and can contain only letters, numbers, and the dash (-) character.
+//	2. Every dash (-) character must be immediately preceded and followed by a letter or number; consecutive dashes are not permitted in container names.
+//	3. All letters in a container name must be lowercase.
+//	4. Container names must be from 3 through 63 characters long.
+//
+// See https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names
+func getValidContainerName(volumeName string) string {
+	containerName := strings.ToLower(volumeName)
+	if len(containerName) > containerNameMaxLength {
+		containerName = containerName[0:containerNameMaxLength]
+	}
+	if !checkContainerNameBeginAndEnd(containerName) || len(containerName) < containerNameMinLength {
+		// now we set as 63 for maximum container name length
+		// todo: get cluster name
+		containerName = k8sutil.GenerateVolumeName("pvc-fuse", uuid.NewUUID().String(), 63)
+		klog.Warningf("the requested volume name (%q) is invalid, so it is regenerated as (%q)", volumeName, containerName)
+	}
+	containerName = strings.Replace(containerName, "--", "-", -1)
+
+	return containerName
+}
+
+func checkContainerNameBeginAndEnd(containerName string) bool {
+	length := len(containerName)
+	if (('a' <= containerName[0] && containerName[0] <= 'z') ||
+		('0' <= containerName[0] && containerName[0] <= '9')) &&
+		(('a' <= containerName[length-1] && containerName[length-1] <= 'z') ||
+			('0' <= containerName[length-1] && containerName[length-1] <= '9')) {
+		return true
+	}
+
+	return false
 }

--- a/pkg/blobfuse/blobfuse_test.go
+++ b/pkg/blobfuse/blobfuse_test.go
@@ -214,3 +214,77 @@ func TestGetStorageAccount(t *testing.T) {
 		}
 	}
 }
+
+func TestGetValidContainerName(t *testing.T) {
+	tests := []struct {
+		volumeName string
+		expected   string
+	}{
+		{
+			volumeName: "aqz",
+			expected:   "aqz",
+		},
+		{
+			volumeName: "029",
+			expected:   "029",
+		},
+		{
+			volumeName: "a--z",
+			expected:   "a-z",
+		},
+		{
+			volumeName: "A2Z",
+			expected:   "a2z",
+		},
+		{
+			volumeName: "1234567891234567891234567891234567891234567891234567891234567891",
+			expected:   "123456789123456789123456789123456789123456789123456789123456789",
+		},
+	}
+
+	for _, test := range tests {
+		result := getValidContainerName(test.volumeName)
+		if !reflect.DeepEqual(result, test.expected) {
+			t.Errorf("input: %q, getValidContainerName result: %q, expected: %q", test.volumeName, result, test.expected)
+		}
+	}
+}
+
+func TestCheckContainerNameBeginAndEnd(t *testing.T) {
+	tests := []struct {
+		containerName string
+		expected      bool
+	}{
+		{
+			containerName: "aqz",
+			expected:      true,
+		},
+		{
+			containerName: "029",
+			expected:      true,
+		},
+		{
+			containerName: "a-9",
+			expected:      true,
+		},
+		{
+			containerName: "0-z",
+			expected:      true,
+		},
+		{
+			containerName: "-1-",
+			expected:      false,
+		},
+		{
+			containerName: ":1p",
+			expected:      false,
+		},
+	}
+
+	for _, test := range tests {
+		result := checkContainerNameBeginAndEnd(test.containerName)
+		if !reflect.DeepEqual(result, test.expected) {
+			t.Errorf("input: %q, checkContainerNameBeginAndEnd result: %v, expected: %v", test.containerName, result, test.expected)
+		}
+	}
+}

--- a/pkg/blobfuse/controllerserver.go
+++ b/pkg/blobfuse/controllerserver.go
@@ -21,13 +21,11 @@ import (
 	"fmt"
 	"strings"
 
-	k8sutil "k8s.io/kubernetes/pkg/volume/util"
-
 	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2018-07-01/storage"
 	azstorage "github.com/Azure/azure-sdk-for-go/storage"
 	"github.com/container-storage-interface/spec/lib/go/csi"
 	"github.com/csi-driver/blobfuse-csi-driver/pkg/util"
-	"github.com/pborman/uuid"
+
 	"k8s.io/klog"
 
 	"google.golang.org/grpc/codes"
@@ -90,9 +88,7 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 	accountName = account
 
 	if containerName == "" {
-		// now we set as 63 for maximum container name length
-		// todo: get cluster name
-		containerName = k8sutil.GenerateVolumeName("pvc-fuse", uuid.NewUUID().String(), 63)
+		containerName = getValidContainerName(name)
 	}
 
 	klog.V(2).Infof("begin to create container(%s) on account(%s) type(%s) rg(%s) location(%s) size(%d)", containerName, accountName, storageAccountType, resourceGroup, location, requestGiB)

